internal/service/auth.go

```diff
@@ -1,61 +1,262 @@
+	tokenIDLength  = 32
 )
 
 type AuthService struct {
 	JWTSecret []byte
+	db        storage.DatabaseStorage
 }
 
 // NewAuthService creates a new authentication service
-func NewAuthService(secret string) *AuthService {
+func NewAuthService(secret string, db storage.DatabaseStorage) *AuthService {
 	return &AuthService{
 		JWTSecret: []byte(secret),
+		db:        db,
 	}
 }
 
-// GenerateToken creates a new JWT token
-func (a *AuthService) GenerateToken() (string, error) {
+// generateTokenID generates a random token ID
+func generateTokenID() (string, error) {
+	b := make([]byte, tokenIDLength)
+	if _, err := rand.Read(b); err != nil {
+		return "", err
+	}
+	return base64.URLEncoding.EncodeToString(b), nil
+}
+
+// GenerateToken creates a new JWT token and stores it in the database
+func (a *AuthService) GenerateToken(publicKey string) (string, error) {
+	ctx := context.Background()
+
+	// Generate a unique token ID
+	tokenID, err := generateTokenID()
+	if err != nil {
+		return "", err
+	}
+
 	expirationTime := time.Now().Add(expireDuration)
 	claims := &Claims{
 		RegisteredClaims: jwt.RegisteredClaims{
 			ExpiresAt: jwt.NewNumericDate(expirationTime),
 			IssuedAt:  jwt.NewNumericDate(time.Now()),
 		},
+		PublicKey: publicKey,
+		TokenID:   tokenID,
 	}
+
+	// Create JWT token
 	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
-	return token.SignedString(a.JWTSecret)
+	tokenString, err := token.SignedString(a.JWTSecret)
+	if err != nil {
+		return "", err
+	}
+
+	// Store token in database
+	tx, err := a.db.Pool().Begin(ctx)
+	if err != nil {
+		return "", err
+	}
+	defer tx.Rollback(ctx)
+
+	_, err = a.db.CreateVaultToken(ctx, types.VaultTokenCreate{
+		PublicKey: publicKey,
+		TokenID:   tokenID,
+		ExpiresAt: expirationTime,
+	})
+	if err != nil {
+		return "", err
+	}
+
+	if err := tx.Commit(ctx); err != nil {
+		return "", err
+	}
+
+	return tokenString, nil
 }
 
-// ValidateToken validates a JWT token
+// ValidateToken validates a JWT token and checks its revocation status
 func (a *AuthService) ValidateToken(tokenStr string) (*Claims, error) {
+	ctx := context.Background()
+
 	claims := &Claims{}
 	token, err := jwt.ParseWithClaims(tokenStr, claims, func(token *jwt.Token) (interface{}, error) {
+		if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
+			return nil, errors.New("unexpected signing method")
+		}
 		return a.JWTSecret, nil
 	})
-	if err != nil || !token.Valid {
+	if err != nil {
+		return nil, errors.New("invalid token: " + err.Error())
+	}
+	if !token.Valid {
 		return nil, errors.New("invalid or expired token")
 	}
+
+	// Validate key fields
+	if claims.PublicKey == "" {
+		return nil, errors.New("token missing public key")
+	}
+	if claims.TokenID == "" {
+		return nil, errors.New("token missing token ID")
+	}
+
+	// Check if token is revoked
+	tx, err := a.db.Pool().Begin(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer tx.Rollback(ctx)
+
+	dbToken, err := a.db.GetVaultToken(ctx, claims.TokenID)
```

Comment:
Also here , given `a.db.GetVaultToken` doesn't need transaction, thus you can remove 126-131,